ambry-store/src/test/java/com.github.ambry.store/IndexTest.java

```diff
@@ -1050,6 +1260,72 @@ private void verifyLogSegmentCount(int expectedCount) {
   }
 
   /**
+   * Sleeps in intervals of 10ms until expected progress is made or {@code maxTimeToCheck} is reached
+   * @param expectedProgress expected progress value
+   * @param maxTimeToCheck max time to check if expected progress is reached
+   * @throws InterruptedException
+   */
+  private void waitUntilExpectedProgress(long expectedProgress, int maxTimeToCheck) throws InterruptedException {
+    int sleptSoFar = 0;
+    while (expectedProgress != index.hardDeleter.getProgress()) {
+      sleptSoFar += 10;
+      Thread.sleep(sleptSoFar);
+      if (sleptSoFar == maxTimeToCheck) {
+        fail("HardDelete failed to catch up in " + maxTimeToCheck);
+      }
+      index.hardDeleter.preLogFlush();
+      index.hardDeleter.postLogFlush();
+    }
+  }
+
+  /**
+   * Verifies that the HardDeleter thread is in the {@code expectedState} for the specified {@code timeToCheckInMs} time.
+   * @param expectedState Expected HardDeleter thread state
+   * @param timeToCheckInMs time for which the state has to be checked for
+   * @param intervalToCheckInMs interval at which the check has to be done
+   */
+  private void verifyHardDeleterThreadState(Thread.State expectedState, long timeToCheckInMs, long intervalToCheckInMs)
+      throws InterruptedException {
+    long timeSoFar = 0;
+    while (timeSoFar < timeToCheckInMs) {
+      assertEquals("Hard Deleter thread status mismatch ", expectedState, index.getHardDeleteThreadStatus());
+      advanceTime(intervalToCheckInMs);
+      timeSoFar += intervalToCheckInMs;
+    }
+  }
+
+  /**
+   * Verify that deleted entries are hard deleted or not based on the {@code hardDeleted} flag
+   * @param deletedIds the set of {@link MockId}s that are deleted
+   * @param hardDeleted {@code true} if entries are expected to be hard deleted, {@code false} otherwise
+   * @throws IOException
+   */
+  private void verifyEntriesForHardDeletes(Set<MockId> deletedIds, boolean hardDeleted) throws IOException {
+    for (MockId id : deletedIds) {
+      IndexValue putValue = allKeys.get(id).getFirst();
+      Offset offset = putValue.getOffset();
+      LogSegment segment = log.getSegment(offset.getName());
+      long size = putValue.getSize() - HARD_DELETE_START_OFFSET - HARD_DELETE_LAST_PART_SIZE;
+      ByteBuffer readBuf = ByteBuffer.allocate((int) size);
+      segment.readInto(readBuf, offset.getOffset() + HARD_DELETE_START_OFFSET);
+      readBuf.flip();
+      if (hardDeleted) {
+        while (readBuf.hasRemaining()) {
+          assertEquals("Hard delete has not zeroed out the data", (byte) 0, readBuf.get());
+        }
+      } else {
+        int zerosCount = 0;
+        while (readBuf.hasRemaining()) {
+          if (Byte.compare(readBuf.get(), (byte) 0) == 0) {
```

Comment:
due to the problem mentioned in [this](https://github.com/linkedin/ambry/pull/521/files#r97720127) comment, this can succeed even if hard delete actually ran.